WIP: blog/content/post/2021-04-18-optiplex_part2.md: initial version of pa… (#412)
3mdeb/news-and-ideas: blog/content/post/2021-04-18-optiplex_part2.md

---
title: Dell OptiPlex and coreboot - a story about porting cursed hardware (part 2)
abstract: 'This post is a second part of the series which tells a story about
porting Dell OptiPlex 9010 workstation to open-source firmware - coreboot. What
was simple is long past us in the first part of the series. This time we are
taking a deeper dive into the technical aspects of Environmental Controller
reverse engineering and locating the problem with machine not booting after AC

> loss.'

Author: Not a part of this PR. We can just copy that to the spreadsheet and work on it there.

> loss.'

Author: @pietrushnic I don't see anything in the repo that would say anything about limiting abstract length. If you have more goals in mind, then maybe write them down please somewhere in README?

> loss.'

Reviewer: @miczyg1 not obvious but:

> post abstract in the header is required for the posts summary in the blog list and must contain from 3 to 5 sentences, please note that abstract would be used for social media and because of that should be focused on keywords/hashtags

https://github.com/3mdeb/news-and-ideas/blob/master/blog/content/post/YYYY-MM-DD-template-post.md

(Elite) or Dell OptiPlex 7010 / 9010. Because our company is a
[licensed coreboot consultant](https://www.coreboot.org/consulting.html),
it would be a shame for us to not have our hardware running coreboot.
While the [HP workstations are supported in coreboot](https://doc.coreboot.org/mainboard/hp/compaq_8200_sff.html)
the Dell machines are not (or rather weren't in 2018 when I started my work
with them). This post will take you into my journey of porting the Dell
OptiPlex 7010/9010 to coreboot. If you did not read the first part, I highly
encourage you to do so [here](https://blog.3mdeb.com/2020/2020-06-24-dell-optiplex-port/).
So let's start were we left last time.

# Super I/O chips

In the previous post I mentioned that one day a power cut occurred and the
machine lost the AC power. When I tried to power the platform on, it was no
longer booting. I kept wondering what has gone wrong. Of course there are lots
of power management related registers out there in the main processor which
could ruin my day when configured incorrectly. So I began looping through them
in the datasheet and looking for issues in the coreboot code, however I could

> only come up with this: https://review.coreboot.org/c/coreboot/+/40347 A minor

Author: ```suggestion
only come up with [this patch](https://review.coreboot.org/c/coreboot/+/40347). A minor
```

No matter how dumb it may sound like, but my builds were always with the
onboard serial port enabled for debugging. Because why should I ever get rid of
my main debugging method?? Of course there are other methods like USB debug
dongle, but what for I have already a serial port... What is more I have
verified it is working (at the beginning I had no idea it stops working after
AC loss).

**Secondly, I have interpreted the symptomes incorrectly**

After the AC loss, the first thing I have noticed is that the led on the power
button is not shining white. For clarification SCH5545 has a two-color LED
driver that allows to change its colors based on the register initialization.
Apparently the OptiPlex 9010 power button is equipped with an orange and white
led. When the white led shines, it means the BIOS has started fetching and
executing the code (which is good). However, if the registers are not
configured correctly the power button was shining in orange color giving some
error signal (the led was shining in intervals). My first conclusion was that
the machine is not even starting after AC loss, because I had the code written

> for the leds to set them white and executed very early.

Author: I don't feel like doing it. No merit in that. They did what they did, but at least I had a satisfying challenge.

something.

Summing up all those issues together I was almost certain it has something to
do with the Super I/O and it prevents the board from booting. Of course the
coreboot had not implemented all functionalities yet, but the bug was so severe
that the board was not booting. I have looked into the datasheet of SCH5627 and
noticed there is an Embedded Memory Interface (EMI) there which implements
access to the Envrionmental Controller internal memory. "That must be it" I
thought...

![That chip...](/img/sch5545_yelling.jpg)

<center>__*It is everything your fault!*__</center>

Typically Super I/O chips are not that complex in design and operation, however
those from SMSC stand out very much. There is no standardized access to
hardware monitoring registers like temperature, fan tachometers and PWM for fan
control (unlike the chips from ITE for example). So basically the manufacturer
left the door wide open to developers and let them design the Environmental
Controller firmware to perform the hardware monitoring and implement any
communication protocol they want through the EMI. That doesn't sound good for

> me since I will have to go through reverse engineering path...

Reviewer: @miczyg1 I would add more arguments that are aligned with EU Directive 2009/24/EC. We should explain why you had no other means of solving this problem? And what was your real target?

Also please be clear that you did that because of interoperability issues with coreboot and reverse engineering was done in your personal unpaid time for the benefit of community.

Author: Understood. Thanks for reference o the Directive.

# Reverse Engineering of the EC

Since I didn't have almost any experience with reverse engineering before and
additionally I was not in favor of UEFI, it was quite hard for me at the
beginning to reverse EFI files. Finding the best working environment is the
first and foremost step to begin the reverse engineering. There are a few
notable frameworks to work with: radare2, Ghidra, IDA and Binary Ninja.

> ![](https://rehex.ninja/img/thumb/decomp-versus.png)

Reviewer: Back in the days I liked this: https://twitter.com/securelyfitz/status/1143213013484232704
Please check your rights to use added graphics.

Author: I was looking for that one but couldn't find it! Great to have that

I decided to give Ghidra a try since it is free and also has some good firmware
utilities for [EFI binaries](https://github.com/al3xtjames/ghidra-firmware-utils)
that were developed by a student on GSoC. Note the tool was developed by NSA,

> so if you have any doubts when using it, consider other tools.

Reviewer: ```suggestion
so if you have any doubts when using it, consider other tools or [Qubes OS disposable VM](https://www.qubes-os.org/doc/disposablevm/) for isolation.
```

The Ghidra firmware utilities are very good to start with, since they parse the
PE header, look for known GUIDs etc. Additionally it provides EDK2 datatypes
integration to name variables, structures and interfaces. For example:

![](/img/ghidra_overview.png)

> AS you can see the window is divided into code listing and the decompiler. As

Reviewer: ```suggestion
As you can see the window is divided into code listing and the decompiler. As
```

automatically. This helps a lot in obtaining some understandable code
initially. Note must be taken since the script translates the types from UEFI
DXE modules (which run from system RAM) and does not recognize PEI modules
(which run from SPI flash directly in cache). So for example I was making a
stupid mistake and reversing PEI modules with
`_ModuleEntryPoint(EFI_HANDLE ImageHandle,EFI_SYSTEM_TABLE *SystemTable)` which
is only true for DXE and UEFI executables. So it is not always good that a
script does something automatically for you.

Long story short I have been reversing those modules one by one which were
related to the SCH5545 and EMI. I spent a lot of time on that and I still have
got zero clues what to do next. This wasn't however wasted time, because I had
to do it sooner or later.

# Final conclusion

Many days passed without any action on my side. Then, one morning I have been
enlightened. "What if the board actually starts but does not give any signs of
life?" and then... Yesss... After the AC loss the Environmental Controller
might lost its configuration and that also means other functions may not work

> as well. Before the AC loss I did nto have to configure Environment Controller

Reviewer: ```suggestion
as well. Before the AC loss, I did not have to configure Environment Controller
```

and fan control. The fan was working silently. However just after the AC was
lost, the fan started to spin at full speed when the board was powered on. It
is possible that the Environmental Controller keeps its configuration as long
as standby power is supplied. So taking it into consideration and all my
previous conclusions I have decided to enable the USB dongle debug instead of

> the onboard serial port. Why? Look at this:

Reviewer: @miczyg1 maybe you can say precisely where below code coming from?

Author: Sure

assumed that the board is not starting at all. This is the first time I have
encountered such issue with Super I/O where the serial port is not working
despite having a known good driver. Just in case I stopped producing builds
with onboard serial port as the main debug method. I bought FT232H which is
supported by coreboot and used it as EHCI USB debug dongle. I was able to get
the debug output from coreboot without worrying about AC loss or whatsoever.

## Summary

So the issue with board not booting after AC loss has been resolved, or at
least located. It took me enormous amount of time to figure it out. But don't
feel sorry, it's a daily firmware developer job to debug for many hours just to
find a simple fix. And I kind of like it. It is a professional deviation I
could say.

In the next post I will disclose more details how I implemented the
Environmental Controller driver for the Dell OptiPlex 9010. It should show some
progress on making the fan control work and how I have achieved the current
implementation level which is available in upstream coreboot repository. Also
you will see how far one has to go to achieve a goal in firmware development.
Stay tuned.

> If you think we can help in improving the security of your firmware or you
> looking for someone who can boost your product by leveraging advanced features
> of used hardware platform, feel free to [book a call with us](https://calendly.com/3mdeb/consulting-remote-meeting)

Reviewer: ```suggestion
If you have a problem with the Environmental Controller (EC) of your platform, looking to
write EC debugging or updating utilities, we are one to discuss. Also, since [EC firmware ecosystem shifts to Zephyr](https://cfp.osfc.io/osfc2020/talk/SXYUQD/) 
and we are huge fans of Zephyr. We would be glad to implement Firmware or even Open Source Firmware for your EC. If you are interested, feel free to [book a call with us](https://calendly.com/3mdeb/consulting-remote-meeting)
```
